pkg/test/main/main.go

```diff
@@ -16,80 +16,205 @@
-		glog.Fatal(err)
+	log.Infof("waiting for the first request...")
+	<-signal
+	log.Infof("initial snapshot %+v", snapshots)
+	log.WithFields(log.Fields{"updates": updates, "requests": requests}).Info("executing sequence")
+
+	for i := 0; i < updates; i++ {
+		snapshots.Version = fmt.Sprintf("v%d", i)
+		log.WithFields(log.Fields{"version": snapshots.Version}).Info("update snapshot")
+		err := config.SetSnapshot(nodeID, snapshots.Generate())
+		if err != nil {
+			log.Errorf("snapshot error %q for %+v", err, snapshots.Generate())
+			os.Exit(1)
+		}
+
+		// pass is true if all requests succeed at least once in a run
+		pass := false
+		for j := 0; j < requests; j++ {
+			ok, failed := callEcho()
+			if failed == 0 && !pass {
+				pass = true
+			}
+			log.WithFields(log.Fields{"batch": j, "ok": ok, "failed": failed, "pass": pass}).Info("request batch")
+			select {
+			case <-time.After(delay):
+			case <-ctx.Done():
+				return
+			}
+		}
+
+		if !pass {
+			log.Errorf("failed all requests in a run %d", i)
+			os.Exit(1)
+		}
+
+		als.Dump(func(s string) { log.Debug(s) })
 	}
 
-	// start envoy
-	envoy := exec.Command("envoy", "-c", bootstrapFile, "--drain-time-s", "1")
-	envoy.Stdout = os.Stdout
-	envoy.Stderr = os.Stderr
-	envoy.Start()
+	log.Info("Test passed!")
+}
+
+// callEcho calls upstream echo service on all listener ports and returns an error
+// if any of the listeners returned an error.
+func callEcho() (int, int) {
+	total := httpListeners + tcpListeners
+	ok, failed := 0, 0
+	ch := make(chan error, total)
+
+	// spawn requests
+	for i := 0; i < total; i++ {
+		go func(i int) {
+			client := http.Client{
+				Timeout: 100 * time.Millisecond,
+			}
+			req, err := client.Get(fmt.Sprintf("http://localhost:%d", basePort+uint(i)))
+			if err != nil {
+				ch <- err
+				return
+			}
+			defer req.Body.Close()
+			body, err := ioutil.ReadAll(req.Body)
+			if err != nil {
+				ch <- err
+				return
+			}
+			if string(body) != test.Hello {
+				ch <- fmt.Errorf("unexpected return %q", string(body))
+				return
+			}
+			ch <- nil
+		}(i)
+	}
 
 	for {
-		if err := test.CheckResponse(listenPort); err != nil {
-			glog.Errorf("ERROR %v", err)
+		out := <-ch
+		if out == nil {
+			ok++
 		} else {
-			glog.Info("OK")
+			failed++
+		}
+		if ok+failed == total {
+			return ok, failed
 		}
+	}
+}
 
-		time.Sleep(1 * time.Second)
+type logger struct{}
+
+func (logger logger) Infof(format string, args ...interface{}) {
+	log.Debugf(format, args...)
+}
+
+type callbacks struct {
+	signal chan struct{}
+}
+
+func (cb *callbacks) OnStreamOpen(id int64, typ string) {
+	log.Debugf("stream %d open for %s", id, typ)
+}
+func (cb *callbacks) OnStreamClosed(id int64) {
+	log.Debugf("stream %d closed", id)
+}
+func (cb *callbacks) OnStreamRequest(int64, *v2.DiscoveryRequest) {
+	if cb.signal != nil {
+		close(cb.signal)
+		cb.signal = nil
+	}
+}
+func (cb *callbacks) OnStreamResponse(int64, *v2.DiscoveryResponse) {}
```

Comment:
Does it makes sense to pass the `Node` here? It can be obtained from the `DiscoveryRequest` in `OnStreamRequest` and `OnFetch`, but without it we have to keep a `map[int64]Node`, which is cumbersome.

Reply:
Yes, it does make sense. This is a quirk of the protocol - typically it's the same node on the stream, but each request comes with a copy of the node.